File: lwaf_3D/ITranslated.java
package lwaf_3D;

import lwaf_core.vec3;

@SuppressWarnings("unused")
public interface ITranslated<R> {
    vec3 getTranslation();
    R setTranslation(vec3 translation);

    default R translateBy(vec3 translation) {
        return setTranslation(getTranslation().add(translation));
    }

    default R setTranslation(float x, float y, float z) {
        return setTranslation(new vec3(x, y, z));
    }

    default R translateBy(float x, float y, float z) {
        return setTranslation(getTranslation().add(new vec3(x, y, z)));
    }
}

File: lwaf_3D/ITranslated.kt
package lwaf_3D

import lwaf_core.plus
import lwaf_core.vec3

interface ITranslated<R> {
    val translation: vec3
    fun setTranslation(translation: vec3): R

    fun translateBy(translation: vec3): R
            = setTranslation(this.translation + translation)

    fun setTranslation(x: Float, y: Float, z: Float): R
             = setTranslation(vec3(x, y, z))

    fun translateBy(x: Float, y: Float, z: Float): R
            = setTranslation(translation + vec3(x, y, z))
}
